modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeClusterServiceFactory.java

```diff
@@ -41,41 +41,51 @@
 
         var cluster = new ClusterImpl()
             .handler(cl -> new ClusterMessageHandler() {
+                /** {@inheritDoc} */
                 @Override public void onMessage(Message message) {
                     messagingService.fireEvent(message);
                 }
 
+                /** {@inheritDoc} */
                 @Override public void onMembershipEvent(MembershipEvent event) {
-                    topologyService.fireEvent(event);
+                    topologyService.onMembershipEvent(event);
                 }
             })
             .config(opts -> opts.memberAlias(context.getName()))
             .transport(opts -> opts.port(context.getPort()))
             .membership(opts -> opts.seedMembers(parseAddresses(context.getMemberAddresses())));
 
         // resolve cyclic dependencies
-        topologyService.setCluster(cluster);
         messagingService.setCluster(cluster);
 
         return new AbstractClusterService(context, topologyService, messagingService) {
+            /** {@inheritDoc} */
             @Override public void start() {
                 cluster.startAwait();
+
+                topologyService.setLocalMember(cluster.member());
```

Comment:
why do you call this method here and not in place of `setCluster`?

Reply:
Cluster is not started hence there is no local member at the moment of the creation